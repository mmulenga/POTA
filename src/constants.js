/**
 * This file contains all string constants for the project
 * Use this instead of hardcoding string values
 * HTML and one-time-use messages are an acceptable exception
 */

/* Conditions */
export const Condition = {
  /* Cardiovascular Diseases */
  AtrialFib: 'Atrial fibrillation / History of irregular heart beat',
  Defib: 'Defibrillator / Pacemaker',
  CAD: 'Coronary artery disease / Cardiac stent',
  CerebralDisease: 'History of cerebral vascular disease / TIA / Stroke',
  VHD: 'Valvular heart disease / Valve Replacement',
  HeartFail: 'Heart failure',
  PVD: 'Peripheral Vascular disease (PVD)',
  /* Pulmonary Diseases */
  PulmonaryVascular: 'History of pulmonary vascular disease',
  PulmDisease: 'Severe COPD, Home Oxygen, Pulmonary HTN',
  /* Other Diseases */
  Risk: 'Age ≥ 69 years old OR ≥ 2 risk factors (HTN, DM, OSA, CKD, BMI > 35)',
  Bleeding: 'Bleeding disorders',
  Anemia: 'History of anemia in past 12 months or ongoing blood loss',
  KidneyDisease: 'Chronic kidney disease / on dialysis',
  Diabetes: 'Diabetes',
  Malignancy: 'Present malignancy / surgery for malignancy',
  Hepatic: 'Hepatic disease',
  Adrenal: 'History of adrenal disease',
  Pituitary: 'History of pituitary disease',
  Endocrine: 'History of major systemic endocrine disease',
  Thyroid: 'Known thyroid disease or on thyroid supplementation',
  Digoxin: 'Digoxin',
  Lithium: 'Lithium',
  Diuretics: 'Diuretics',
  ACEI: 'ACE-I',
  ARB: 'ARB',
  NSAIDS: 'NSAIDS',
  Anticoagulant: 'Anticoagulants',
  Antiplatelet: 'Antiplatelet (ASA excluded)',
  Steroid: 'Systemic steroid use within 6 months',
};

/* Medical Examinations */
export const Exam = {
  GnS: 'Group and Screen',
  ECG: 'ECG',
  CBC: 'CBC',
  RenPanel: 'Renal Panel (Creat + Lytes)',
  PTTINR: 'PTT/INR',
  LFT: 'LFT',
  Gluc: 'Gluc',
  HbA1C: 'HbA1C',
  CXR: 'CXR',
  TSH: 'TSH',
};

/* Glossary Entries */
export const Glossary = {
  AtrialFib: 'No extra information.',
  Defib: 'No extra information.',
  CAD: 'No extra information.',
  CardiacStent: 'No extra information.',
  CerebralDisease: 'No extra information.',
  PulmonaryVascular: 'No extra information.',
  VHD: 'No extra information.',
  HeartFail: 'No extra information.',
  PVD: 'No extra information.',
  /* Pulmonary Diseases */
  PulmDisease: '• Spirometric values AND symptomatic assessment should be sought. \n' +
               '• Post Bronchodilator FEV1: 30 - 50% of predicted; FEV1/FVC < 0.70 \n' +
               '• History of exacerbations (including prior hospitalizations) should be recorded.\n' +
               '• Worsening air flow limitation, greater SOB, reduced exercise capacity, having an ' +
               'impact on quality of life, and ability to complete activities of daily living should be present. \n',
  /* Other Diseases */
  Risk: 'No extra information.',
  Bleeding: '• Hemophiliac. \n' +
            '• History of extensive investigation due to a bleeding episode. \n' +
            '• History of extensive bleeding needing transfusion in previous surgeries. \n' +
            '• Bruising easily more than usual. \n',
  Anemia: 'No extra information.',
  ActiveBleeding: 'No extra information.',
  KidneyDisease: 'No extra information.',
  Renal: '(AKI). \n' +
         '• Chronic kidney disease / on dialysis. \n' +
         '• Diabetes. \n' +
         '• Heart failure. \n' +
         '• Age ≥ 65. \n' +
         '• Liver disease. \n' +
         '• Use of drugs with nephrotoxic potential in the perioperative period (in particular NSAIDs after surgery). \n',
  Diabetes: 'No extra information.',
  Malignancy: 'No extra information.',
  Hepatic: '• Alcohol abuse (i.e., female: 14 drinks / week; male: 21 drinks / week). \n' +
           '• Jaundice due to liver disease. \n' +
           '• All forms of hepatitis. \n' +
           '• Cirrhosis; Hepatic cancer. \n',
  Adrenal: 'No extra information.',
  Pituitary: '• Acromegaly. \n' +
             '• Cushing’s disease. \n' +
             '• Sheehan syndrome. \n' +
             '• Growth hormone deficiency. \n' +
             '• Hyper/hypo pituitarism. \n',
  Endocrine: '(Excluding Diabetes) \n' +
             '• Parathyroid disease \n' +
             '• Addison’s disease \n' +
             '• Cushing’s syndrome \n' +
             '• Graves’ disease \n',
  Thyroid: 'No extra information.',
  /* Medications */
  Digoxin: 'No extra information.',
  Lithium: 'No extra information.',
  Diuretics: 'Furosemide; Hydrochlorothiazide; Acetazolamide; Spironolactone; Chlorthalidone',
  ACEI: 'Captopril; Enalapril; Ramipril; Trandolapril; Perindopril',
  ARB: 'Valsartan; Candesartan; Irbesartan; Telmisartan; Losartan; Olmesartan',
  NSAIDS: 'No extra information.',
  Anticoagulant: 'Warfarin; Heparins; Dabigatran; Apixaban; Rivaroxaban',
  Antiplatelet: 'Clopidogrel; Ticagrelor; Prasugrel; Aspirin/Dipyridamole (Aggrenox)',
  Steroid: 'Prednisone; Prednisolone; Methylprednisolone; Hydrocortisone; Dexamethasone; Oral Budesonide',
};
