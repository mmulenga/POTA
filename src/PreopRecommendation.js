// Singleton data, not a class
import '@/configuration';
import { GetExams, GetConditionalExams, ConditionalExam, GetExamValidity } from '@/tag';
import { Exam } from '@/constants';

/**
 * ExamSummary is a simple data struct with two properties:
 * @param exams {String[]}: a list of exams to be performed
 * @param conditionalExams {ConditionalExam[]} a list of ConditionalExams to be optionally performed
 */
export class ExamSummary {
  constructor(exams, conditionalExams) {
    this.exams = exams;
    this.conditionalExams = conditionalExams;
  }
}

/**
 * Given a list of patient conditions, return an ExamSummary object
 * The ConditionalExams returned will be prefiltered so as to not include any
 * already required exams or any redundant conditional phrases (i.e. no new exams)
 * @param {String[]} patientConditions
 * @returns {ExamSummary} Summary of examinations to be performed
 */
export function PatientExamsNeeded(patientConditions) {
  const examAggregation = new Set();
  const conditionalExamAggregation = [];

  for (let i = 0; i < patientConditions.length; i += 1) {
    const exams = GetExams(patientConditions[i]);
    for (let j = 0; j < exams.length; j += 1) {
      examAggregation.add(exams[j]);
    }
  }

  // conditionPhrasesUsed is used to dedupe conditional phrases that are equivalent
  // It may come up anytime that a single row of the grid is broken up into Tags
  const conditionPhrasesUsed = [];

  for (let i = 0; i < patientConditions.length; i += 1) {
    // Filter out any previously used condition phrases
    const conditionalExams = GetConditionalExams(patientConditions[i])
      .filter(c => !(conditionPhrasesUsed.includes(c.conditionPhrase)));
    // Add remaining phrases to conditionPhrasesUsed
    conditionPhrasesUsed.push(...conditionalExams.map(c => c.conditionPhrase));

    for (let j = 0; j < conditionalExams.length; j += 1) {
      // Make a new copy of the ConditionalExam with any preconsidered exams filtered out
      const ce = new ConditionalExam(
        conditionalExams[j].conditionPhrase,
        conditionalExams[j].exams.filter(e => !(examAggregation.has(e))),
      );
      // Only aggregate this conditionalExam if it is not entirely preconsidered
      if (ce.exams.length > 0) {
        conditionalExamAggregation.push(ce);
      }
    }
  }

  // Unconditionally perform a Group and Screen
  return new ExamSummary([Exam.GnS, ...examAggregation], conditionalExamAggregation);
}

/**
 * Get the validity period for a given exam
 * @param {String} exam The name of the exam in question
 */
export function ExamValidity(exam) {
  return GetExamValidity(exam);
}
