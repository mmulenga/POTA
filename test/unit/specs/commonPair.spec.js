import { Condition, Exam } from '@/constants';
import { PatientExamsNeeded } from '@/PreopRecommendation';

let requiredExams;
let conditionalExams;
let result;

/**
 * Testing PatientExamsNeeded() with test cases given by the stakeholders
 * which represent very common sets of comorbidities that a patient may have.
 * All of the test cases are in the test matrix.
 */
describe('Testing PatientExamsNeeded() with common pairs of comorbidities', () => {
  it('returns correct set of preop exams for AtrialFib + Age > 69', () => {
    result = PatientExamsNeeded([Condition.AtrialFib, Condition.Risk]);
    requiredExams = [Exam.GnS, Exam.ECG];
    conditionalExams = [];
    expect(result.exams).toEqual(requiredExams);
    expect(result.conditionalExams).toEqual(conditionalExams);
  });
  it('set of preop exams for AtrialFib + Age > 69 to match snapshot', () => {
    expect(result.exams).toMatchSnapshot();
    expect(result.conditionalExams).toMatchSnapshot();
  });

  it('returns correct set of preop exams for CAD + CardiacStent', () => {
    result = PatientExamsNeeded([Condition.AtrialFib, Condition.CardiacStent]);
    requiredExams = [Exam.GnS, Exam.ECG];
    conditionalExams = [];
    expect(result.exams).toEqual(requiredExams);
    expect(result.conditionalExams).toEqual(conditionalExams);
  });

  it('set of preop exams for CAD + CardiacStent to match snapshot', () => {
    expect(result.exams).toMatchSnapshot();
    expect(result.conditionalExams).toMatchSnapshot();
  });


  it('returns correct set of preop exams for AtrialFib + VHD', () => {
    result = PatientExamsNeeded([Condition.AtrialFib, Condition.VHD]);
    requiredExams = [Exam.GnS, Exam.ECG];
    conditionalExams = [Exam.CBC];
    expect(result.exams).toEqual(requiredExams);
  });
  it('set of preop exams for AtrialFib + VHD to match snapshot', () => {
    expect(result.exams).toMatchSnapshot();
  });

  it('returns correct set of preop exams for HeartFail + VHD', () => {
    result = PatientExamsNeeded([Condition.HeartFail, Condition.VHD]);
    requiredExams = [Exam.GnS, Exam.ECG, Exam.RenPanel];
    conditionalExams = [];
    expect(result.exams).toEqual(requiredExams);
    expect(result.conditionalExams).toEqual(conditionalExams);
  });
  it('set of preop exams for HeartFail + VHD to match snapshot', () => {
    expect(result.exams).toMatchSnapshot();
    expect(result.conditionalExams).toMatchSnapshot();
  });

  it('returns correct set of preop exams for CAD + Diabetes', () => {
    result = PatientExamsNeeded([Condition.CAD, Condition.Diabetes]);
    requiredExams = [Exam.GnS, Exam.ECG, Exam.RenPanel, Exam.Gluc, Exam.HbA1C];
    conditionalExams = [Exam.HbA1C];
    expect(result.exams).toEqual(requiredExams);
  });
  it('set of preop exams for CAD + Diabetes to match snapshot', () => {
    expect(result.exams).toMatchSnapshot();
  });

  it('returns correct set of preop exams for Diabetes + Malignancy', () => {
    result = PatientExamsNeeded([Condition.Diabetes, Condition.Malignancy]);
    requiredExams = [Exam.GnS, Exam.ECG_diabetes, Exam.RenPanel, Exam.Gluc, Exam.HbA1C, Exam.CBC, Exam.CXR];
    expect(result.exams).toEqual(requiredExams);
    expect(result.conditionalExams[0].exams).toEqual([Exam.ECG, Exam.PTTINR]);
  });
  it('set of preop exams for Diabetes + Malignancy to match snapshot', () => {
    expect(result.exams).toMatchSnapshot();
    expect(result.conditionalExams[0].exams).toMatchSnapshot();
  });

  it('returns correct set of preop exams for AtrialFib + Anticoagulant', () => {
    result = PatientExamsNeeded([Condition.AtrialFib, Condition.Anticoagulant]);
    requiredExams = [Exam.GnS, Exam.ECG, Exam.CBC, Exam.RenPanel];
    conditionalExams = [Exam.PTTINR];
    expect(result.exams).toEqual(requiredExams);
    expect(result.conditionalExams[0].exams).toEqual(conditionalExams);
  });
  it('set of preop exams for Atrial Fibrillation + Anticoagulants to match snapshot', () => {
    expect(result.exams).toMatchSnapshot();
    expect(result.conditionalExams[0].exams).toMatchSnapshot();
  });

  it('returns correct set of preop exams for 2 >= risks + ACEI', () => {
    result = PatientExamsNeeded([Condition.Risk, Condition.ACEI]);
    requiredExams = [Exam.GnS, Exam.ECG, Exam.RenPanel];
    conditionalExams = [];
    expect(result.exams).toEqual(requiredExams);
    expect(result.conditionalExams).toEqual(conditionalExams);
  });
  it('set of preop exams for 2 >= risks + ACEI to match snapshot', () => {
    expect(result.exams).toMatchSnapshot();
    expect(result.conditionalExams).toMatchSnapshot();
  });

  it('returns correct set of preop exams for AtrialFib + Digoxin', () => {
    result = PatientExamsNeeded([Condition.AtrialFib, Condition.Digoxin]);
    requiredExams = [Exam.GnS, Exam.ECG, Exam.RenPanel];
    conditionalExams = [];
    expect(result.exams).toEqual(requiredExams);
    expect(result.conditionalExams).toEqual(conditionalExams);
  });
  it('set of preop exams for Defibrillator + Use of Digoxin to match snapshot', () => {
    expect(result.exams).toMatchSnapshot();
    expect(result.conditionalExams).toMatchSnapshot();
  });

  it('returns correct set of preop exams for Defib + Digoxin', () => {
    result = PatientExamsNeeded([Condition.Defib, Condition.Digoxin]);
    requiredExams = [Exam.GnS, Exam.ECG, Exam.RenPanel];
    conditionalExams = [];
    expect(result.exams).toEqual(requiredExams);
    expect(result.conditionalExams).toEqual(conditionalExams);
  });
  it('set of preop exams for Defib + Digoxin to match snapshot', () => {
    expect(result.exams).toMatchSnapshot();
    expect(result.conditionalExams).toMatchSnapshot();
  });

  it('returns correct set of preop exams for Adrenal + Steroid', () => {
    result = PatientExamsNeeded([Condition.Adrenal, Condition.Steroid]);
    requiredExams = [Exam.GnS, Exam.RenPanel, Exam.Gluc];
    conditionalExams = [];
    expect(result.exams).toEqual(requiredExams);
    expect(result.conditionalExams).toEqual(conditionalExams);
  });
  it('set of preop exams for Adrenal + Steroid to match snapshot', () => {
    expect(result.exams).toMatchSnapshot();
    expect(result.conditionalExams).toMatchSnapshot();
  });

  it('returns correct set of preop exams for Bleeding + Anemia', () => {
    result = PatientExamsNeeded([Condition.Bleeding, Condition.Anemia]);
    requiredExams = [Exam.GnS, Exam.CBC, Exam.PTTINR];
    conditionalExams = [];
    expect(result.exams).toEqual(requiredExams);
    expect(result.conditionalExams).toEqual(conditionalExams);
  });
  it('set of preop exams for Bleeding + Anemia to match snapshot', () => {
    expect(result.exams).toMatchSnapshot();
    expect(result.conditionalExams).toMatchSnapshot();
  });

  it('returns correct set of preop exams for KidneyDisease + Diabetes', () => {
    result = PatientExamsNeeded([Condition.KidneyDisease, Condition.Diabetes]);
    requiredExams = [Exam.GnS, Exam.CBC, Exam.RenPanel, Exam.ECG_diabetes, Exam.Gluc, Exam.HbA1C];
    conditionalExams = [Exam.HbA1C];
    expect(result.exams).toEqual(requiredExams);
  });
  it('set of preop exams for KidneyDisease + Diabetes to match snapshot', () => {
    expect(result.exams).toMatchSnapshot();
  });
});
