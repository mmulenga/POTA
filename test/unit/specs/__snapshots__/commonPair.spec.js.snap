// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing PatientExamsNeeded() with common pairs of comorbidities set of preop exams for 2 >= risks + ACEI to match snapshot 1`] = `
Array [
  "Group and Screen",
  "ECG",
  "Renal Panel (Creat + Lytes)",
]
`;

exports[`Testing PatientExamsNeeded() with common pairs of comorbidities set of preop exams for 2 >= risks + ACEI to match snapshot 2`] = `Array []`;

exports[`Testing PatientExamsNeeded() with common pairs of comorbidities set of preop exams for Adrenal + Steroid to match snapshot 1`] = `
Array [
  "Group and Screen",
  "Renal Panel (Creat + Lytes)",
  "Gluc",
]
`;

exports[`Testing PatientExamsNeeded() with common pairs of comorbidities set of preop exams for Adrenal + Steroid to match snapshot 2`] = `Array []`;

exports[`Testing PatientExamsNeeded() with common pairs of comorbidities set of preop exams for Atrial Fibrillation + Anticoagulants to match snapshot 1`] = `
Array [
  "Group and Screen",
  "ECG",
  "CBC",
  "Renal Panel (Creat + Lytes)",
]
`;

exports[`Testing PatientExamsNeeded() with common pairs of comorbidities set of preop exams for Atrial Fibrillation + Anticoagulants to match snapshot 2`] = `
Array [
  "PTT/INR",
]
`;

exports[`Testing PatientExamsNeeded() with common pairs of comorbidities set of preop exams for AtrialFib + Age > 69 to match snapshot 1`] = `
Array [
  "Group and Screen",
  "ECG",
]
`;

exports[`Testing PatientExamsNeeded() with common pairs of comorbidities set of preop exams for AtrialFib + Age > 69 to match snapshot 2`] = `Array []`;

exports[`Testing PatientExamsNeeded() with common pairs of comorbidities set of preop exams for AtrialFib + VHD to match snapshot 1`] = `
Array [
  "Group and Screen",
  "ECG",
]
`;

exports[`Testing PatientExamsNeeded() with common pairs of comorbidities set of preop exams for Bleeding + Anemia to match snapshot 1`] = `
Array [
  "Group and Screen",
  "CBC",
  "PTT/INR",
]
`;

exports[`Testing PatientExamsNeeded() with common pairs of comorbidities set of preop exams for Bleeding + Anemia to match snapshot 2`] = `Array []`;

exports[`Testing PatientExamsNeeded() with common pairs of comorbidities set of preop exams for CAD + CardiacStent to match snapshot 1`] = `
Array [
  "Group and Screen",
  "ECG",
]
`;

exports[`Testing PatientExamsNeeded() with common pairs of comorbidities set of preop exams for CAD + CardiacStent to match snapshot 2`] = `Array []`;

exports[`Testing PatientExamsNeeded() with common pairs of comorbidities set of preop exams for CAD + Diabetes to match snapshot 1`] = `
Array [
  "Group and Screen",
  "ECG",
  "Renal Panel (Creat + Lytes)",
  "Gluc",
  "HbA1C",
]
`;

exports[`Testing PatientExamsNeeded() with common pairs of comorbidities set of preop exams for Defib + Digoxin to match snapshot 1`] = `
Array [
  "Group and Screen",
  "ECG",
  "Renal Panel (Creat + Lytes)",
]
`;

exports[`Testing PatientExamsNeeded() with common pairs of comorbidities set of preop exams for Defib + Digoxin to match snapshot 2`] = `Array []`;

exports[`Testing PatientExamsNeeded() with common pairs of comorbidities set of preop exams for Defibrillator + Use of Digoxin to match snapshot 1`] = `
Array [
  "Group and Screen",
  "ECG",
  "Renal Panel (Creat + Lytes)",
]
`;

exports[`Testing PatientExamsNeeded() with common pairs of comorbidities set of preop exams for Defibrillator + Use of Digoxin to match snapshot 2`] = `Array []`;

exports[`Testing PatientExamsNeeded() with common pairs of comorbidities set of preop exams for Diabetes + Malignancy to match snapshot 1`] = `
Array [
  "Group and Screen",
  "ECG",
  "Renal Panel (Creat + Lytes)",
  "Gluc",
  "HbA1C",
  "CBC",
  "CXR",
]
`;

exports[`Testing PatientExamsNeeded() with common pairs of comorbidities set of preop exams for Diabetes + Malignancy to match snapshot 2`] = `
Array [
  "PTT/INR",
]
`;

exports[`Testing PatientExamsNeeded() with common pairs of comorbidities set of preop exams for HeartFail + VHD to match snapshot 1`] = `
Array [
  "Group and Screen",
  "ECG",
  "Renal Panel (Creat + Lytes)",
]
`;

exports[`Testing PatientExamsNeeded() with common pairs of comorbidities set of preop exams for HeartFail + VHD to match snapshot 2`] = `Array []`;

exports[`Testing PatientExamsNeeded() with common pairs of comorbidities set of preop exams for KidneyDisease + Diabetes to match snapshot 1`] = `
Array [
  "Group and Screen",
  "CBC",
  "Renal Panel (Creat + Lytes)",
  "ECG",
  "Gluc",
  "HbA1C",
]
`;
